Error generating code for const ref subscript operator

Add to `Common.h` TestIndexedProperties:

``` cpp
    int operator[](const int& key);
```

Error:

```
../gen/Common/Common.cs(4932,24): error CS0213: You cannot use the fixed statement to take the address of an already fixed expression
../gen/Common/Common.cs(4932,46): error CS0266: Cannot implicitly convert type `int**' to `int*'. An explicit conversion exists (are you missing a cast?)
```

C# code:

``` csharp
        public int this[int* key]
        {
            get
            {
                fixed (int* __refParamPtr0 = &key)
                {
                    var arg0 = __refParamPtr0;
                    var __ret = Internal.OperatorSubscript_9((__Instance + __PointerAdjustment), arg0);
                    return __ret;
                }
            }
        }
```

--- a/src/Generator/Generators/CSharp/CSharpMarshal.cs
+++ b/src/Generator/Generators/CSharp/CSharpMarshal.cs
@@ -245,7 +245,7 @@ namespace CppSharp.Generators.CSharp
 
                 Context.Before.WriteLine("var {0} = {1};", ptrName,
                     Context.ReturnVarName);
-                
+
                 var res = $"{ptrName} == IntPtr.Zero? null : ({typedef})Marshal.GetDelegateForFunctionPointer({ptrName}, typeof({typedef}))";
                 Context.Return.Write(res);
                 return true;
@@ -492,7 +492,8 @@ namespace CppSharp.Generators.CSharp
 
             var pointee = pointer.Pointee.Desugar();
             if (Context.Function != null && pointer.IsPrimitiveTypeConvertibleToRef() &&
-                Context.MarshalKind != MarshalKind.VTableReturnValue)
+                Context.MarshalKind != MarshalKind.VTableReturnValue &&
+                Context.Function.OperatorKind != CXXOperatorKind.Subscript)
             {
                 var refParamPtr = string.Format("__refParamPtr{0}", Context.ParameterIndex);
                 var templateSubstitution = pointer.Pointee as TemplateParameterSubstitutionType;
@@ -580,13 +581,17 @@ namespace CppSharp.Generators.CSharp
                 if (isRefParam)
                 {
                     var typeName = Type.TypePrinterDelegate(finalPointee);
-
-                    if (param.IsInOut)
-                        Context.Before.WriteLine("{0} _{1} = {1};", typeName, param.Name);
+                    if (Context.Function.OperatorKind == CXXOperatorKind.Subscript)
+                        Context.Return.Write(param.Name);
                     else
-                        Context.Before.WriteLine("{0} _{1};", typeName, param.Name);
+                    {
+                        if (param.IsInOut)
+                            Context.Before.WriteLine($"{typeName} _{param.Name} = {param.Name};");
+                        else
+                            Context.Before.WriteLine($"{typeName} _{param.Name};");
 
-                    Context.Return.Write("&_{0}", param.Name);
+                        Context.Return.Write($"&_{param.Name}");
+                    }
                 }
                 else
                 {

--- a/tests/CSharp/CSharp.Tests.cs
+++ b/tests/CSharp/CSharp.Tests.cs
@@ -834,4 +834,14 @@ public unsafe class CSharpTests : GeneratorTestFixture
             complexArrayElement.Dispose();
         }
     }
+
+    [Test]
+    public void TestConstRefIndexer()
+    {
+        using (var indexproperty = new TestIndexedProperties())
+        {
+            int a = 2;
+            Assert.That(indexproperty[&a], Is.EqualTo(2));
+        }
+    }
 }

--- a/tests/CSharp/CSharp.h
+++ b/tests/CSharp/CSharp.h
@@ -1193,3 +1193,14 @@ struct HasComplexArray
 {
     ComplexArrayElement complexArray[ARRAY_LENGTH_MACRO];
 };
+
+class DLL_API TestIndexedProperties
+{
+public:
+    int operator[](const int& key);
+};
+
+int TestIndexedProperties::operator[](const int& key)
+{
+    return key;
+}

